jupyter_server/serverapp.py

```diff
@@ -1487,31 +1490,96 @@ def init_server_extension_config(self):
+                                    "`load_jupyter_server_extension` function "
+                                    "was found.".format(module_name=module_name)
+                                )
+                                self.log.warning(log_msg)
+                            else:
+                                self.enabled_extensions[extloc] = extmod
+                        else:
+                            log_msg = _(
+                                "{module_name} is missing critical metadata. Check "
+                                "that _jupyter_server_extension_paths returns `app` "
+                                "and/or `module` as keys".format(module_name=module_name)
+                            )
+                            self.log.warn(log_msg)
+                except KeyError:
+                    # A KeyError suggests that the module does not have a
+                    # _jupyter_server_extension-path.
+                    log_msg = _(
+                        "Error loading server extensions in "
+                        "{module_name} module. There is no `_jupyter_server_extension_path` "
+                        "defined at the root of the extension module. Check "
+                        "with the author of the extension to ensure this function "
+                        "is added.".format(module_name=module_name)
+                    )
+                    self.log.warning(log_msg)
+                except:
+                    if self.reraise_server_extension_failures:
+                        raise
+                    self.log.warning(_("Error loading server extension module: %s"), module_name,
+                                  exc_info=True)
+
+    def load_server_extensions(self):
         """Load any extensions specified by config.
 
         Import the module, then call the load_jupyter_server_extension function,
         if one exists.
 
         The extension API is experimental, and may change in future releases.
         """
-        # Initialize extensions
-        for modulename, enabled in sorted(self.jpserver_extensions.items()):
-            if enabled:
-                try:
-                    mod = importlib.import_module(modulename)
-                    func = getattr(mod, 'load_jupyter_server_extension', None)
-                    if func is not None:
-                        func(self)
-                        # Add debug log for loaded extensions.
-                        self.log.debug("%s is enabled and loaded." % modulename)
-                    else:
-                        self.log.warning("%s is enabled but no `load_jupyter_server_extension` function was found" % modulename)
-                except Exception:
-                    if self.reraise_server_extension_failures:
-                        raise
-                    self.log.warning(_("Error loading server extension %s"), modulename,
-                                  exc_info=True)
```

Comment:
Should we not keep this exception guard? Or is its implementation moved? I.e. shouldn't the actual calling of the `load_jupyter_server_extension` function be guarded?

Reply:
Depends on what you mean. It *is* guarded in that, if it no load_jupyter_server_extension function is found, the serverapp doesn't crash but throws (more helpful) console log warnings. This PR improves the overall messaging around server extension failures.

Reply:
I meant: Previously, this code had a `try: load_jupyter_server_extension(...) except Exception`. Now it has:
```python
            try:
                # This function was prefixed with an underscore in in v1.0
                # because this shouldn't be a public API for most extensions.
                func = getattr(extension, '_load_jupyter_server_extension')
                func(self)
                self.log.debug(log_msg)
            except AttributeError:
                # For backwards compatibility, we will still look for non
                # underscored loading functions.
                func = getattr(extension, 'load_jupyter_server_extension')
                func(self)
                self.log.debug(log_msg)
            except:
                warn_msg = _(
                    "{extkey} is enabled but no "
                    "`_load_jupyter_server_extension` function "
                    "was found.".format(extkey=extkey)
                )
                self.log.warning(warn_msg)
```

So if the `load_jupyter_server_extension` throws an error (`AttributeError` or not), the error message will be less informative.